llvm/tools/llvm-readobj/ELFDumper.cpp

```diff
@@ -7365,6 +7367,19 @@ template <class ELFT> void LLVMELFDumper<ELFT>::printDynamicTable() {
   W.startLine() << "]\n";
 }
 
+template <class ELFT> void JSONELFDumper<ELFT>::printDynamicTable() {
+  Elf_Dyn_Range Table = this->dynamic_table();
+  ListScope L(this->W, "DynamicSection");
+  for (const auto &Entry : Table) {
+    DictScope D(this->W);
+    uintX_t Tag = Entry.getTag();
+    std::string Value = this->getDynamicEntry(Tag, Entry.getVal());
+    this->W.printHex("Tag", Tag);
+    this->W.printString("Value", Value);
+    this->W.printString("Type", this->Obj.getDynamicTagAsString(Tag));
```

Comment:
I think we need to give this output a bit more consideration, given that this is JSON and therefore intended to be machine-readable (otherwise, we could just use LLVM-style output). In particular, I have the following two thoughts:
1) Tag/Type fields are just different ways of representing the same information. Do we need both of them?
2) The Value field is a string, yet this is almost never the "natural" representation of the value for a dynamic tag. It also is rarely the actual value, because it is an interpretation of that value. In my opinion, this potentially deserves splitting into two fields: a raw integer value (which is just the "bytes on disk" numeric value for the entry), and an interpreted value, in a format that is appropriate for the interpretation, e.g. a list of strings, or an integer (in the latter case, arguably the interpreted field may not be necessary).

Thoughts?

Reply:
1. I do like having the type string there but I can appreciate that it's duplicative.
2. Yeah. what about a construct like
```
"Value": { "bytes": 16 },
"Type": "PLTRELSZ",
},
....
"Value": { "Shared library": "[D]" }
"Type": "NEEDED"
},
...
"Value": {"value": 4096 },
"Type": "HASH"
},
```

and so on?

Reply:
I'm not opposed to having the type field, if you think it's useful for readability of the output.

As I understand JSON, I'm not sure we need a nested object for the Value field, because the individual DynamicTable entries can have different types for that field. In other words, I think it could look something like this:

```
...
"Value": 16
"Type": "PLTRELSZ"
...
"Value": ["A", "B"]
"RawValue":4321
"Type": "NEEDED"
...
"Value": 4096
"Type": "HASH"
```
Note that I've added "RawValue" for the case where the value isn't just the raw number in the dynamic table. For some dynamic entries, the value is the actual meaning. In other cases, it's a pointer to some other information stored somewhere. The "bytes" bit of the size is unnecessary, since the size values are implicitly in bytes.

A possible variation on my above suggestion is the following, where every entry has a "Value" member, which is the raw value stored in the dynamic table, and in the cases where that value has some interpreted meaning, there's an additional entry. In this way, a generic parser doesn't need to understand the tag types to be able to record the raw values, since all the "Value" entries have the same type.
```
"Type": "PLTRELSZ",
"Value": 16
...
"Type": "NEEDED",
"Value": 1234,
"Libraries": ["A", "B", "C"]
...
"Type": "SONAME",
"Value": 100,
"Name": "my-library"
```
(For simplicity in both examples, I've not included separate Tag/Type fields, but these can be included as per the current behaviour)

Reply:
I went with option 2. How does that strike you?

Reply:
Once we are good with the format I'll clean up the commit message so the output matches @ilovepi to keep everything tidy

Reply:
Sorry, forgot to respond to this: current format is basically fine now, thanks. I'll continue reviewing and make any more comments I find.